python/ray/autoscaler/_private/resource_demand_scheduler.py

```diff
@@ -118,6 +139,12 @@ def get_nodes_to_launch(
         logger.info("Resource demands: {}".format(resource_demands))
         logger.info("Unfulfilled demands: {}".format(unfulfilled))
         max_to_add = self.max_workers - sum(node_type_counts.values())
+        if resource_requests:
+            nodes_to_add_based_on_requests = get_nodes_for(
+                self.node_types, node_type_counts, max_to_add,
+                resource_requests)
+        else:
+            nodes_to_add_based_on_requests = {}
         nodes_to_add_based_on_demand = get_nodes_for(
             self.node_types, node_type_counts, max_to_add, unfulfilled)
```

Comment:
I am not sure if this is commutative. Shouldn't we instead do:

```
        if resource_requests:
            nodes_to_add = get_nodes_for(
                self.node_types, node_type_counts, max_to_add,
                resource_requests+unfulfilled)
        else:
            nodes_to_add = get_nodes_for(
                self.node_types, node_type_counts, max_to_add,
                unfulfilled)```

Reply:
No, we need to calculate both separately in order to avoid the upscaling constraint for requested.

Reply:
Then lets do:
```
        if resource_requests:
            nodes_to_add_based_on_requests = get_nodes_for(
                self.node_types, node_type_counts, max_to_add,
                unfulfilled)
            total_nodes_to_add = get_nodes_for(
                self.node_types, node_type_counts, max_to_add,
                resource_requests+unfulfilled)
        else:
            nodes_to_add_based_on_requests = {}
            total_nodes_to_add = get_nodes_for(
                self.node_types, node_type_counts, max_to_add,
                unfulfilled)
```
Otherwise, I think you will waste resources. I think the operation is not commutative. If the resource_requests is requesting 1 GPU, and the unfulfilled is 1 CPU, you might end up launching 2 nodes when a single node can be sufficient to satisfy both.

Reply:
I'm confused, isn't this exactly the same as the current code? Because if resource_requests is empty dict, the two branches do the same thing. when calculating total_nodes_to_add.

Reply:
I see, you add them in the unfulfilled in advance, thanks for clarifying!